File: 백준/Gold/2589. 보물섬/보물섬.cc
#include <bits/stdc++.h>
using namespace std;

int n, m, ans, visited[50][50], dy[] = { 0, 0, 1, -1 }, dx[] = { 1, -1, 0, 0 };
char a[50][50];

void bfs(int y, int x) {
	queue<pair<int, int>> q;
	visited[y][x] = 1;
	q.push({ y, x });

	while (q.size()) {
		tie(y, x) = q.front();
		q.pop();

		for (int i = 0; i < 4; i++) {
			int ny = y + dy[i];
			int nx = x + dx[i];

			if (ny >= 0 && ny < n && nx >= 0 && nx < m && a[ny][nx] == 'L' && !visited[ny][nx]) {
				visited[ny][nx] = visited[y][x] + 1;
				q.push({ ny, nx });
			}
		}
	}
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n >> m;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++)
			cin >> a[i][j];
	}

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			if (a[i][j] == 'L') {
				memset(visited, 0, sizeof(visited));
				bfs(i, j);
				for (int k = 0; k < n; k++) {
					for (int l = 0; l < m; l++)
						ans = max(ans, visited[k][l]);
				}
			}
		}
	}

	cout << ans - 1;

	return 0;
}

File: 백준/Gold/2589. 보물섬/보물섬.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, m, ans = 0;
	static char[][] a;
	static List<Node> land = new ArrayList<>();
	static int[] dy = {-1, 0, 1, 0};
	static int[] dx = {0, 1, 0, -1};

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());

		a = new char[n][m];
		for (int i = 0; i < n; i++) {
			String s = br.readLine();
			for (int j = 0; j < m; j++) {
				a[i][j] = s.charAt(j);
				if (a[i][j] == 'L') {
					land.add(new Node(i, j));
				}
			}
		}

		for (Node node : land) {
			bfs(node);
		}

		bw.write((ans - 1) + "");

		bw.flush();
	}

	static void bfs(Node node) {
		Queue<Node> q = new LinkedList<>();
		int[][] visited = new int[n][m];
		q.add(node);
		visited[node.y][node.x] = 1;

		while (!q.isEmpty()) {
			Node cur = q.poll();
			for (int i = 0; i < 4; i++) {
				int ny = cur.y + dy[i];
				int nx = cur.x + dx[i];
				if (ny < 0 || nx < 0 || ny >= n || nx >= m)
					continue;
				if (a[ny][nx] == 'W' || visited[ny][nx] != 0)
					continue;
				q.add(new Node(ny, nx));
				visited[ny][nx] = visited[cur.y][cur.x] + 1;
			}
		}

		int ret = 0;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				ret = Math.max(ret, visited[i][j]);
			}
		}

		ans = Math.max(ans, ret);
	}

	static class Node {
		int y, x;

		Node(int y, int x) {
			this.y = y;
			this.x = x;
		}
	}
}
